### ISSUE TITLE
pam_limits appears to read limits off PID 1 and use those, and really shouldn't do that.

### ISSUE THREAD

**Issue by poettering** [linux-pam/linux-pam#85]
It appears current pam_limits reads the limits in effect on PID 1 and sets those in some cases for any login session. This is highly problematic, since what PID 1 internally does for these limits might be appropriate for PID 1 itself, but not at all for other processes. In particular if systemd is PID 1 it bumps RLIMIT_NOFILE to very high values, since it might need large numbers of fds (in particular as new kernel APIs are designed with the idea that fds are cheap, see cgroups/bpf apis and stuff). it doesn't pass these high limits on to the children it forks off though, it uses much more conservative values there.

This causes issue like https://github.com/systemd/systemd/issues/10921

Quite frankly any logic to read resource limits off PID 1 to set them for PAM sessions is a really bad idea and should be removed entirely. They are in no way suitable defaults but should be considered internal implementation details of PID 1.

**Comment by keszybz** [comment]
The code in question is at https://github.com/linux-pam/linux-pam/blob/master/modules/pam_limits/pam_limits.c#L381. The intent is clearly to read "kernel limits", but of course there's no guarantee whatsoever that PID 1 has the same limits as the kernel. So this depends on the internal implementation details of the init process.

**Comment by t8m** [comment]
Fortunately the easy workaround is to not use the set_all option.

**Comment by keszybz** [comment]
Yes. But under Debian, pam_limits.so behaves as if set_all was set. They have a slightly older version (1.1.8-3.8), but I don't see any relevant changes in the upstream code that would affect this.

**Comment by t8m** [comment]
This must be some Debian-specific change. The upstream code definitely doesn't do this without the set_all option.

**Comment by keszybz** [comment]
It is. @mbiebl figured it out [https://github.com/systemd/systemd/issues/10921#issuecomment-442511099]:
> https://sources.debian.org/src/pam/1.1.8-3.8/debian/patches-applied/027_pam_limits_better_init_allow_explicit_root/#L66

> This patch removes the if (ctrl & PAM_SET_ALL) { check.
So Debian's pam_limits.so behaves as set_all is always set.

**Comment by poettering** [comment]
I figure that means it's Debian's job to fix this. That said I am still sure the concept should be removed upstream too, since it doesn't make much sense to read the internal limits of PID 1...

**Comment by t8m** [comment]
Well there might be other init systems than systemd which do not adjust the limits. But we should certainly at least document that the option does not work right with systemd as init.

**Comment by poettering** [comment]
> Well there might be other init systems than systemd which do not adjust the limits. But we should certainly at least document that the option does not work right with systemd as init.

i think it would be wiser to assume a default of "rlimits are internal to init systems", rather than to assume a default of "rlimits are not touched and public" but then consider systemd to be the outlier.

**Comment by smcv** [comment]
> That said I am still sure the concept should be removed upstream too, since it doesn't make much sense to read the internal limits of PID 1...

I'm revisiting this to try to get some movement in Debian's PAM packaging to arrange for more reasonable rlimits to be set. The original `RLIMIT_NOFILE` issue has been avoided by another Debian-specific patch, but it still has unintended effects for other rlimits that systemd adjusts on itself, such as `RLIMIT_MEMLOCK`.

Am I correct to think that what PAM *really* wants to read is "the kernel's default rlimits" (which are non-obvious because some are kernel-version-dependent and some are RAM-dependent); but because there is no kernel API to achieve that, the closest achievable thing is to read pid 1's limits, and hope that pid 1 did not change them?

@poettering, would it be feasible for systemd to record the inheritable state that it received from the kernel (which AIUI cannot be discovered programmatically any other way?) and export it somewhere, perhaps in `/run`? (I assume it already records that state in-memory somewhere so that it can pass it down the process hierarchy to system services?)

**Comment by t8m** [comment]
> Am I correct to think that what PAM _really_ wants to read is "the kernel's default rlimits" (which are non-obvious because some are kernel-version-dependent and some are RAM-dependent); but because there is no kernel API to achieve that, the closest achievable thing is to read pid 1's limits, and hope that pid 1 did not change them?

Yes, that is the purpose of the `set_all` code in pam_limits.

**Comment by keszybz** [comment]
I'll try to answer this in lieu of @poettering:

The default limits are exposed over dbus, see https://www.freedesktop.org/software/systemd/man/org.freedesktop.systemd1.html, DefaultLimitCPU, etc.

The concept of "kernel defaults" is not well defined, because even PID1 doesn't know if it is getting the kernel default or if previous code running as PID1 already modified the limits. (This "previous code" could be another instance of systemd, or maybe `/bin/bash` if somebody used `init=/bin/bash`, or maybe a process in the initrd, or a container manager, and so on.) So the best that systemd can do is say "whatever we inherited" + some adjustments that are appropriate for user processes, which is what the dbus properties expose.

I don't think this makes much sense to expose anything more. Systemd goes to a lot of effort to set various resource limits for children, and it doesn't make sense to reset a subset of limits from pam. If one wants a really "clean slate", they should use systemd APIs to spawn a new unit afresh. This will be more comprehensive, it is more flexible, more debuggable, and in general better. It doesn't make sense for us to add another API to query limits to allow resetting them from user processes, since that API would be systemd-specific anyway, and under systemd a different approach is better.

Of course non-systemd systems are not covered by this, and if you want to keep `set_all` for them, then I would suggest putting a big `if (!sd_booted())` guard around the implementation so that it continues working in that case without breaking things for the majority of users (who are on systemd).

I'm not quite sure why Debian keeps carrying those patches which are at best useless to 99% of it's users, but the solution with `if (!sd_booted())` would at least avoid breakage.

**Comment by poettering** [comment]
> Am I correct to think that what PAM _really_ wants to read is "the kernel's default rlimits" (which are non-obvious because some are kernel-version-dependent and some are RAM-dependent); but because there is no kernel API to achieve that, the closest achievable thing is to read pid 1's limits, and hope that pid 1 did not change them?

Yes, this appears to be the intention.

> @poettering, would it be feasible for systemd to record the inheritable state that it received from the kernel (which AIUI cannot be discovered programmatically any other way?) and export it somewhere, perhaps in `/run`? (I assume it already records that state in-memory somewhere so that it can pass it down the process hierarchy to system services?)

I am not convinced this is really a good idea. I disagree with the intention here: i.e. blanekt replacing the limits the PAM services (hopefully carefully) set up for their PAM sessions wildcard with kernel defaults sounds like a really bad idea. I mean, various of the kernel limits are quite frankly really bad choices. Most prominently we nowadays override the RLIMIT_NOFILE hard limit to really high values, with the explicit intention this is inherited down all userspace. We made this after discussing this with relevant kernel people who clarified to us that fds are cheap now and thus the low limit unnecessary (and in fact some new kernel apis are already designed with fds being cheap now in mind, i.e. are quite wasteful with fds) — but they couldn't easily change the built-in kernel defaults so easily for compat. In systemd we pass each service a cleaned up, well-defined, stable execution environment, with the intention that these limits remain in effect. If you now stick pam_limits into all PAM stacks with a configuration that resets things to the kernel's defaults (for whatever that means) then you basically tape over all this, and break the clear intention of limits encoded in the unit files.

Hence: pam_limits is great, and I think it should be in all PAM stacks, but it should be a noop by default — only if the user decides to actually configure something in its configuration file it should do something, but blanket resetting things to defaults is a *bad* *idea*, both because the defaults are bad, and because this breaks how the system is designed these days with a focus on running everything in clean execution environments and intention for settings to propagate down the tree.

Or in other words: doing blanket clean-ups of resource limits via pam_rlimits is a sysvinit thing really, where there was no clean execution environment for system services, and thus there was some value in cleaning things up. It is also a workaround for PAM services that were sloppy in figuring out which limits to set, reset or pass down.

**Comment by smcv** [comment]
Thanks, this is really useful information.

I think there are two related issues here:

* This upstream issue: the `set_all` implementation is wrong with systemd, because it makes the assumption that the limits that are active on PID 1 are an appropriate default;
* A Debian-specific downstream issue: the patched PAM in Debian is using the `set_all` behaviour even in situations where it should not

> I'm not quite sure why Debian keeps carrying those patches [...] a sysvinit thing really, where there was no clean execution environment for system services

[The rationale given for the ancestor of the patch used in Debian](https://bugs.debian.org/63230), back in 2000 (!), is not about system services as I previously thought: it's about `su -` wanting to undo any limits inherited from the caller of `su`, and replace them with "something reasonable". This seems like a situation where `set_all` might indeed have had some value, if there is a way for `set_all` to find out what a reasonable limit would have looked like.

If I was the maintainer of PAM in Debian (I'm not) and I was asked to make `pam_limits` default to the `set_all` behaviour to benefit `su`, my response would have been that `/etc/pam.d/su` can easily specify `pam_limits.so set_all` if it wants to, and therefore `pam_limits.so` should continue to have its upstream behaviour. However, that still requires `set_all` to be implementable.

If privilege-escalation mechanisms like `su -` want to be able to reset the limits to "something reasonable", that does seem like a reasonable feature to have. At the moment, `pam_limits` tries to get its "reasonable" values by parsing `/proc/1/limits` - but as noted here, that doesn't give the right answers for systemd, or for any other non-trivial init system that might need to adjust its own rlimits independent of the rlimits that will be the default further down the process hierarchy. It also assumes that the kernel's default rlimits for pid 1 are "reasonable", but we've seen that they are not always - modern systems can (and sometimes must) handle a lot more than 4096 file descriptors per process.

**Comment by poettering** [comment]
Quite frankly, if "su" doesn't sanitize the process rlimits when changing privs, then this is clearly a bug in "su". If your job is to do privilege isolation between parent and child processes and you are not sanitizing rlimits then that's a big problem, because the limits are relevant for security choices. I mean, some even very explicitly: RLIMIT_RTPRIO is about allowing unpriv processes to acquire rt privs, if left untouched during priv switching, which is a great way to freeze the machine. 

Seriously, if "su" doesn't clean up rlimits on its own, then that's a bug in "su", and pam_rlimits is a terrible work-around for that. Please file a bug against "su" to sanitize the rlimits when changing privs, if this really is an issue, still.

**Comment by smcv** [comment]
> Quite frankly, if "su" doesn't sanitize the process rlimits when changing privs, then this is clearly a bug in "su".

I agree, but this brings us back to the problem of what "sanitize" means.

One possible implementation of resetting the limits is to use a PAM stack with `pam_limits.so set_all`, which is exactly resetting all the rlimits to *something* - but then we're back in this issue with `pam_limits.so`.

Or, `su` implementations could reset the limits to some default internally, instead of relying on PAM to know - but if they do that, they need to know what reasonable rlimits look like, which presents them with the same problem that PAM has. They can't read limits from pid 1, because that would be wrong for systemd; they can't read defaults from the kernel, because it doesn't tell us what its defaults are, except by setting them when it runs the initial init process; and hard-coding limits in the `su` implementation doesn't seem great either, because some of the kernel's defaults are context-dependent (scale up with amount of RAM), and hard-coding them in `su` means there's one more place to change if their arbitrary values turn out to be inappropriate.

In practice, all the `su`-like utilities seem to rely on PAM to be their implementation of this, if anything. There are probably more of them than you think...

util-linux `su`, used in most distributions (including Debian since 2018), doesn't appear to alter limits itself: it delegates all that to PAM.

shadow `su`, used in Debian prior to 2018, also doesn't appear to alter limits itself: it delegates all that to PAM.

GNU coreutils `su` was removed from the source package in 2018 so I haven't checked it further.

If I'm reading correctly, `sudo` doesn't seem to alter limits for the child process. It temporarily elevates various limits and disables core dumps while running its own privileged code, but puts back the previous limits before executing user code.

`pkexec` doesn't do anything special with limits; neither do `cado`, `chiark-really` or `doas`. `calife` and `super` only touch resource limits so that they can temporarily disable core dumps (a subset of what `sudo` does.

The PAM stacks are, unfortunately, distro-specific. As a couple of examples of well-known distributions, it looks as though Fedora has `pam_limits.so` in the `system-auth` substack, without `set_all`; util-linux `su` and `su-l` inherit that from `system-auth`, but neither uses `set_all`. Debian does not have `pam_limits.so` in the `common-*` files that are our equivalent of Fedora `system-auth`; many individual services such as `su` and `su-l` use `pam_limits.so`, without an explicit `set_all` (although Debian-specific patches to PAM make it behave in a way that is equivalent to `set_all`).

**Comment by poettering** [comment]
@karelzak i think util-linux `su` tool should start sanitizing rlimits if it doesn't already. any take on that? i.e. in particular RLIMIT_RTPRIO and RLIMIT_NICE must be sanitized I am sure.

**Comment by karelzak** [comment]
> @karelzak i think util-linux `su` tool should start sanitizing rlimits if it doesn't already. any take on that? i.e. in particular RLIMIT_RTPRIO and RLIMIT_NICE must be sanitized I am sure.

OK, how? ;-) I think @smcv's previous note about it is correct, it's difficult to say what the default is. 

Anyway, see can use zero for RLIMIT_NICE and RLIMIT_RTPRIO if we want to hard code it to su(1).

**Comment by karelzak** [comment]
OK, I have improved su/runuser. Now it calls setrlimit() before it opens a new PAM session, so pam_limits.so is still possible to use to set something else.
